## Main snakemake for domain analysis on protein data
## Date : Decembre 2024
## Authors :


# Import python modules
# ---------------------
import os
import json

# Function to load JSON files
# ---------------------------
def load_json(file_path):
    with open(file_path, "r") as file:
        return json.load(file)

# Assign environment variables
# ----------------------------
globals().update(load_json("../environment_path.json"))


# Configuration
# -------------
configfile: "analyse.json"
configfile: "assemblies.json"


# List of assemblies
# ------------------
ACCESSNB = config["assembly_list"]

# Name of the analyse. The name is  used  pathGTDriftResource:
# it contains the resources used by the analyse (the fasta alignment files)
# and generated by the analysis (the hmm profiles and databases)
# The directories are structured as follows
# ANALYSE_DIR/
# ...........reference_alignments/ (should be present at start)
# .............................. DOMAIN1/
# ...................................... Reference_alignement_DOMAIN1.fst
# .......................................[Paralogous_alignement_DOMAIN1.fst](if DOMAIN1 is selected for paralogous analysis) 
# .......................................[Paralogous_alignement_DOMAIN1.fst](if DOMAIN1 is selected for paralogous analysis) 
# ...............................DOMAIN2/
# .......................................Reference_alignement_DOMAIN2.fst
# .......................................[Paralogous_alignement_DOMAIN2.fst](if DOMAIN2 is selected for paralogous analysis) 
# .......................................[Paralogous_alignement_DOMAIN2.fst](if DOMAIN2 is selected for paralogous analysis) 
# ...........hmm_profiles/ (will be generated by the pipeline)
# ........................DOMAIN1/
# ............................... Reference_alignement_DOMAIN1.hmm
# ................................[Paralogous_alignement_DOMAIN1.hmm](if DOMAIN1 is selected for paralogous analysis) 
# ................................[Paralogous_alignement_DOMAIN1.hmm](if DOMAIN1 is selected for paralogous analysis) 
# ........................DOMAIN2/
# ................................Reference_alignement_DOMAIN2.hmm
# ............................... [Paralogous_alignement_DOMAIN2.hmm](if DOMAIN2 is selected for paralogous analysis) 
# ................................[Paralogous_alignement_DOMAIN2.hmm](if DOMAIN2 is selected for paralogous analysis) 
# ...........hmm_databases/ (will be generated by the pipeline)
# ........................[database_DOMAIN1.hmm{.h3i,.h3f,.h3m.h3p}] (if DOMAIN1 is selected for paralogous analysis)
# ........................[database_DOMAIN2.hmm{.h3i,.h3f,.h3m.h3p}] (if DOMAIN2 is selected for paralogous analysis)
# -------------------------------------------------------------------------------------------------------------------
ANALYSE_DIR  = config["analyse_dir_name"]

# The reference alignment of each domain 
# ---------------------------------------
DOMAIN_REFERENCES = config["domain_references"]

# List of domains to be fully processed. 
# --------------------------------------
DOMAINS = config["domains"]

# List of domains to be processed without paralogy checks. 
# --------------------------------------------------------
DOMAINS_SIMPLE = config["domains_simple"]

# Name of global results directory.
# The directory is located in pathGTDriftGlobalResults
# ----------------------------------------------------
GLOBAL_RESULTS = config["analyse_dir_name"]

# Name of genome specific results directory.
# The directory is located in genome_assembly/{accession}/analyses/
# -----------------------------------------------------------------
GENOME_RESULTS = config["analyse_dir_name"]

# config for hmm
# --------------
if config["mode"] == "guix":
    RUNCMD = "guix shell hmmer -- "
else:
    RUNCMD = ""
    
    
    
# function to get the name of the reference alignment for a domain
# ----------------------------------------------------------------
def get_domain(wildcards):
    domain = wildcards.domain
    return domain 
      
# function to get the name of the reference alignment for a domain
# ----------------------------------------------------------------
def get_reference(wildcards):
    fname = DOMAIN_REFERENCES.get(wildcards.domain, "")
    return fname 

# function to get the path of the reference alignment for a domain
# ----------------------------------------------------------------
def get_reference_file(wildcards):
    fname = DOMAIN_REFERENCES.get(wildcards.domain, "")
    domain = wildcards.domain
    return pathGTDriftResource + ANALYSE_DIR + "hmm_profiles/"+ domain+"/"+fname+".hmm"

# get the files and directories describing the reference alignments
# -----------------------------------------------------------------   
directories, files = glob_wildcards(pathGTDriftResource + ANALYSE_DIR + "reference_alignments/{dir}/{file}.fst")

# Rules
# -----

# -----------------------------------------------
# all : inputs define the to be files generated . 
# -----------------------------------------------
rule all:
    """
    Generate the candidates in fasta format
    """
    input:  
        # Candidates after paralog checks
        # --------------------------------
        candidates=expand(
            pathGTDriftData
            + "genome_assembly/{accession}/analyses/" + GENOME_RESULTS
            + "candidates_{domain}.fasta",
            accession=ACCESSNB,domain=DOMAINS
        ),

        # Candidates without paralog checks
        # --------------------------------
        candidates_simple=expand(
            pathGTDriftData
           + "genome_assembly/{accession}/analyses/" + GENOME_RESULTS
           + "candidates_1_{domain}.fasta",
            accession=ACCESSNB,domain=DOMAINS_SIMPLE
        ),
        # Summaries without paralog checks
        # --------------------------------        
        summary_simple=expand(
            pathGTDriftData
            + "genome_assembly/{accession}/analyses/" + GENOME_RESULTS 
            + "summary_hmmsearch_{accession}_{domain}.csv",
            accession=ACCESSNB,domain=DOMAINS_SIMPLE),
            
         # Whole analyse summary
         # ---------------------
         whole_summary=expand(
            pathGTDriftData
            + "genome_assembly/{accession}/analyses/" + GENOME_RESULTS 
            + "whole_summary.csv",
            accession=ACCESSNB),
            
         # Concatenation of assemblies results
         # -----------------------------------
         concat_assemblies=pathGTDriftGlobalResults + GLOBAL_RESULTS + "results.csv"  
        
            

# Modules snakemake
# -----------------

include: "../utils/module_stats_domain.smk"
include: "../utils/module_check_paralogs.smk"
include: "../utils/module_hmm.smk"
include: "../utils/module_concatenate.smk"

